zerver/lib/actions.py

```diff
@@ -817,6 +817,67 @@ def do_send_messages(messages):
     # intermingle sending zephyr messages with other messages.
     return already_sent_ids + [message['message'].id for message in messages]
 
+def do_send_typing_notification(notification):
+    # type: (Dict[str, Any]) -> None
+    if notification['recipient'].type == Recipient.PERSONAL:
+        notification['recipients'] = [notification['sender'],
+                                      get_user_profile_by_id(notification['recipient'].type_id)]
+    elif notification['recipient'].type == Recipient.HUDDLE:
+        # We use select_related()/only() here, while the PERSONAL case above uses
+        # get_user_profile_by_id() to get UserProfile objects from cache.  Streams will
+        # typically have more recipients than PMs, so get_user_profile_by_id() would be
+        # a bit more expensive here, given that we need to hit the DB anyway and only
+        # care about the email from the user profile.
+        fields = [
+            'user_profile__id',
+            'user_profile__email',
+            'user_profile__is_active',
+            'user_profile__realm__domain'
+        ]
+        query = Subscription.objects.select_related("user_profile", "user_profile__realm").only(*fields).filter(
+                recipient=notification['recipient'], active=True)
+        notification['recipients'] = [s.user_profile for s in query]
+    else:
+        raise ValueError('Bad recipient type')
```

Comment:
It's possible that you want to extract this piece of logic out of `do_send_messages` into a function and just call that function (with an added check forbidding Recipient.STREAM).